src/script/api/script_road.cpp

```diff
@@ -356,57 +382,57 @@ static int32_t LookupWithBuildOnSlopes(::Slope slope, const Array<> &existing, i
  */
-static bool NormaliseTileOffset(int32_t *tile)
+static std::optional<RoadPartOrientation> ToRoadPartOrientation(const TileIndex &tile)
 {
-		if (*tile == 1 || *tile == -1) return true;
-		if (*tile == ::TileDiffXY(0, -1)) {
-			*tile = -2;
-			return true;
-		}
-		if (*tile == ::TileDiffXY(0, 1)) {
-			*tile = 2;
-			return true;
-		}
-		return false;
+	if (tile == ScriptMap::GetTileIndex(0, -1)) return RoadPartOrientation::NW;
+	if (tile == ScriptMap::GetTileIndex(1, 0)) return RoadPartOrientation::SW;
+	if (tile == ScriptMap::GetTileIndex(0, 1)) return RoadPartOrientation::SE;
+	if (tile == ScriptMap::GetTileIndex(-1, 0)) return RoadPartOrientation::NE;
+	return std::nullopt;
 }
 
-/* static */ SQInteger ScriptRoad::CanBuildConnectedRoadParts(ScriptTile::Slope slope_, Array<> &&existing, TileIndex start_, TileIndex end_)
+/* static */ SQInteger ScriptRoad::CanBuildConnectedRoadParts(ScriptTile::Slope slope_, Array<TileIndex> &&existing, TileIndex start, TileIndex end)
 {
 	::Slope slope = (::Slope)slope_;
-	int32_t start = start_.base();
-	int32_t end = end_.base();
 
 	/* The start tile and end tile cannot be the same tile either. */
 	if (start == end) return -1;
 
-	for (size_t i = 0; i < existing.size(); i++) {
-		if (!NormaliseTileOffset(&existing[i])) return -1;
+	std::vector<RoadPartOrientation> existing_orientations;
+	existing_orientations.reserve(existing.size());
+	for (const auto &t : existing) {
+		auto orientation = ToRoadPartOrientation(t);
+		if (!orientation) return -1;
+		existing_orientations.push_back(*orientation);
 	}
 
-	if (!NormaliseTileOffset(&start)) return -1;
-	if (!NormaliseTileOffset(&end)) return -1;
+	auto start_orientation = ToRoadPartOrientation(start);
+	auto end_orientation = ToRoadPartOrientation(end);
+	if (!start_orientation || !end_orientation) return -1;
 
 	/* Without build on slopes the characteristics are vastly different, so use
 	 * a different helper function (one that is much simpler). */
-	return _settings_game.construction.build_on_slopes ? LookupWithBuildOnSlopes(slope, existing, start, end) : LookupWithoutBuildOnSlopes(slope, existing, start, end);
+	return _settings_game.construction.build_on_slopes ?
+			LookupWithBuildOnSlopes(slope, existing_orientations, *start_orientation, *end_orientation) :
+			LookupWithoutBuildOnSlopes(slope, existing_orientations, *start_orientation, *end_orientation);
 }
 
 /* static */ SQInteger ScriptRoad::CanBuildConnectedRoadPartsHere(TileIndex tile, TileIndex start, TileIndex end)
 {
 	if (!::IsValidTile(tile) || !::IsValidTile(start) || !::IsValidTile(end)) return -1;
 	if (::DistanceManhattan(tile, start) != 1 || ::DistanceManhattan(tile, end) != 1) return -1;
 
-	/*                                           ROAD_NW              ROAD_SW             ROAD_SE             ROAD_NE */
-	const TileIndexDiff neighbours[] = {::TileDiffXY(0, -1), ::TileDiffXY(1, 0), ::TileDiffXY(0, 1), ::TileDiffXY(-1, 0)};
+	const TileIndex neighbours[] = {
```

Comment:
Does this need to be in the same order as the RoadPartOrientation enum?

Reply:
No, I specifically left the order here the same so the observable behaviour will remain the same.
That it later gets mapped to a different numerical value in the enumeration does not affect the observable logic.